Add support for PPC and PPC64LE dynamic recompilers

dreamer opened the pull request:
This PR takes over from #427 

Same initial work from jmarsh + other patches, that were posted in Vogons thread to bring PPC64LE online + tiny change to enable dynrec on PPC by default.

kcgen commented:
Will give this PR a round on my PPC 7400 and report back.  Not sure how we can test PPC64 though?

dreamer commented:
> Not sure how we can test PPC64 though?

I don't think we can and probably we won't find anyone with access to physical ppc64le before merging this… but considering it's a new feature (ppc64le is not supported right now at all) - merging this won't be a regression, so I am ok with addressing the issues post 0.76.0 :)

Let's just make sure these changes work correctly on PPC (+after quick reading through the files I've seen some areas worthy of small cleanup).

Right now I am mostly concerned about bringing all pending changes in CPU area, so I could continue with cleanups of cache for fixing SELinux / memory protection stuff.

We already know these changes build on ppc64le (not sure about warning numbers though, I will need to repeat builds).

Long-term… I think we need to implement a special "batch" mode for testing - in this mode emulator would start normally, but no SDL window would be created (at all) and all output going to the DOS shell should be either logged to stdout or to a file.

dreamer commented:
Ah, I don't need to repeat builds, after enabling dynrec ppc64le arch has +13 warnings (39 vs 52).

New warnings:
```
core_dynrec/../../fpu/fpu_instructions.h:145:49: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
core_dynrec/cache.h:503:107: warning: format '%d' expects argument of type 'int', but argument 2 has type 'Bitu {aka long unsigned int}' [-Wformat=]
core_dynrec/cache.h:504:84: warning: format '%d' expects argument of type 'int', but argument 2 has type 'Bitu {aka long unsigned int}' [-Wformat=]
core_dynrec/cache.h:504:84: warning: format '%d' expects argument of type 'int', but argument 3 has type 'Bitu {aka long unsigned int}' [-Wformat=]
core_dynrec/decoder_basic.h:722:38: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
core_dynrec/decoder_basic.h:733:38: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
core_dynrec/decoder_basic.h:811:13: warning: 'void dyn_lea_mem_mem(HostReg, void*, void*, Bitu, Bits)' defined but not used [-Wunused-function]
core_dynrec/decoder_opcodes.h:1143:9: warning: enumeration value 'LOOP_JCXZ' not handled in switch [-Wswitch]
core_dynrec/decoder_opcodes.h:1143:9: warning: enumeration value 'LOOP_NONE' not handled in switch [-Wswitch]
core_dynrec/dyn_fpu.h:390:91: warning: format '%d' expects argument of type 'int', but argument 2 has type 'Bitu {aka long unsigned int}' [-Wformat=]
core_dynrec/dyn_fpu.h:390:91: warning: format '%d' expects argument of type 'int', but argument 3 has type 'Bitu {aka long unsigned int}' [-Wformat=]
core_dynrec/risc_ppc64le.h:148:92: warning: overflow in implicit constant conversion [-Woverflow]
core_dynrec/risc_ppc64le.h:270:18: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
```

kcgen commented:
Update works perfectly on my 32-bit PPC, including enabling DYNREC via `./configure`.

dreamer commented:
> Confirming this PR is working on my 32-bit PPC, including enabling DYNREC via `./configure`.

@kcgen This draft includes ac28d83725ad776ad80fd9b622587ca725be6d2e - can you caonfirm, that explicit enabling of dynrec during configure script is not needed?

kcgen commented:
> > Confirming this PR is working on my 32-bit PPC, including enabling DYNREC via `./configure`.
> 
> @kcgen This draft includes [ac28d83](https://github.com/dosbox-staging/dosbox-staging/commit/ac28d83725ad776ad80fd9b622587ca725be6d2e) - can you confirm, that explicit enabling of dynrec during configure script is not needed?

``` text
uname -a
Linux pillow 5.5.7-ofast-powermac-sawtooth-rev8-gcc-9.2.1 #1 Mon Mar 2 17:25:09 PST 2020 ppc ppc ppc GNU/Linux

po/ppc-dynrec-ported-2 ✓]:.../src/dosbox-staging $ ./configure 
...
checking whether byte ordering is bigendian... yes
checking whether memory handling in the CPU Core will be inlined... yes
checking for target cpu type... PowerPC
checking whether Apple user wants to override the build process to produce a 32 bit binary... no, not on Apple
checking whether x86 dynamic cpu core will be enabled... no
checking whether recompiling cpu core will be enabled... yes
checking whether fpu emulation will be enabled... yes
checking whether the x86/x64 assembly fpu core will be enabled... no
checking whether to enable unaligned memory access... yes
...

po/ppc-dynrec-ported-2 ✓]:.../src/dosbox-staging $ grep DYN config.h
/* #undef C_DYNAMIC_X86 */
#define C_DYNREC 1
```

dreamer commented:
@kcgen Thanks! :)

*Note for myself for later about removal of warnings in decoder_basic.h:722 and 733: https://godbolt.org/z/in35uV*

kcgen commented:
> @kcgen Thanks! :)
> 
> _Note for myself for later about removal of warnings in decoder_basic.h:722 and 733: https://godbolt.org/z/in35uV_

```
// Can directly assign if the target is an actual type and not another char* array

uintptr_t new_solution2(uint8_t *tlb_addr) {
    core_dynrec.readdata = host_readw(tlb_addr);
    return core_dynrec.readdata * core_dynrec.readdata;
}

// if we're assigning into unaligned char* memory, then we'd use:
// write_unaligned_uint16(target_mem, host_readw(tlb_addr);
```
See direct assignment here: https://godbolt.org/z/o6acpW

Results comparing the three output values: https://onlinegdb.com/S1eKEHtp8

dreamer commented:
> ```
> // We can directly assign the result to a variable if it's a type big
> // enough to the hold return value, and not another unaligned char* array.
> //
> uintptr_t new_solution2(uint8_t *tlb_addr) {
>     core_dynrec.readdata = host_readw(tlb_addr);
>     return core_dynrec.readdata * core_dynrec.readdata;
> }
>```

Uh, of course - I missed the forest because of all the trees :) Thank you!

When I wrote `memcpy` initially I was under the impression that `core_dynrec.readdata` is a buffer (only later I checked and discovered it is a single Bitu guaranteed to be aligned). I just need to double-check if the result is exactly the same as original "aliasing-by-pointer"  regardless of endianness (I want to be sure exactly the same bytes are being written to Bitu - I have a weird hunch, that in big-endian assignment is equivalent, but in little-endian it's not (or maybe the other way around?)).

dreamer commented:
Updated list of new warnings after recent changes (previous +13 down to +9):
```
core_dynrec/../../fpu/fpu_instructions.h:145:49: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
core_dynrec/decoder_basic.h:722:38: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
core_dynrec/decoder_basic.h:733:38: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
core_dynrec/decoder_basic.h:831:13: warning: 'void dyn_lea_mem_mem(HostReg, void*, void*, Bitu, Bits)' defined but not used [-Wunused-function]
core_dynrec/decoder_basic.h:832:23: warning: attribute ignored [-Wattributes]
core_dynrec/dyn_fpu.h:390:91: warning: format '%d' expects argument of type 'int', but argument 2 has type 'Bitu {aka long unsigned int}' [-Wformat=]
core_dynrec/dyn_fpu.h:390:91: warning: format '%d' expects argument of type 'int', but argument 3 has type 'Bitu {aka long unsigned int}' [-Wformat=]
core_dynrec/risc_ppc64le.h:127:92: warning: overflow in implicit constant conversion [-Woverflow]
core_dynrec/risc_ppc64le.h:240:18: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
```

kcgen commented on src/cpu/core_dynrec/dyn_fpu.h:
```diff
@@ -387,11 +389,12 @@ static void dyn_fpu_esc3(){
 			break;
 		default:
-			LOG(LOG_FPU,LOG_WARN)("ESC 3:Unhandled group %d subfunction %d",decode.modrm.reg,decode.modrm.rm);
+			FPU_LOG_WARN(3, false, decode.modrm.reg, decode.modrm.rm);
```
Should the 2nd arg be true like the others below? (Just a question)

dreamer replied:
Nope:
`false` -> `"ESC 3:…"`
`true` -> `"ESC 3 EA:…"`

I have no idea what "EA" means, but that's how upstream designed interface of FPU_LOG_WARN helper function…

kcgen replied:
Ah! Subtle; I didn't catch that - thank for pointing out this difference.

dreamer commented:
Progress: last 3 warnings introduced by this PR are:
```
core_dynrec/decoder_basic.h:722:38: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
core_dynrec/decoder_basic.h:733:38: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
core_dynrec/risc_ppc64le.h:127:92: warning: overflow in implicit constant conversion [-Woverflow]
```
… except "implicit constant conversion" turned out to be happening in 14 places (our warning counter turned it into 1 warning because warning points to a macro.

kcgen commented:
> dereferencing type-punned pointer will break strict-aliasing

Ripe for the `*_unaligned_*()` functions!
_(not for this PR, but in general when we clean up the aliasing across the board)._

dreamer commented:
Time to mark this as "ready for review". The amount of code in here is huuge - I won't claim I went through it in extreme detail (I don't have complete understanding on dynrec cores yet).

The functionality was implemented in first 3 commits; all small changes afterwards are cleanups and filling-in the gaps or fixing dynrec-related compiler warnings. The last commit exists only to trigger ppc64le builds, won't be merged to master branch.

There are 2 warning fixes, that I decided to drop for now (I'll explain in details later).

@kcgen I think this qualifies for one last test round using your PPC machine :)

@kas1e Can you give it a go as well? This branch includes buildsystem changes, so you shouldn't need to manually enable dynrec any more. Also, I believe you had issues with some games (Command and Conquer IIRC?). Are they still an issue? (I think I found one bug in floating point emulation on big-endian machines, but the fix is not included in this branch).

dreamer commented:
Overall, this PR lowers the number of warnings (e.g. -14 warnings on x86+debug+dynrec build configuration), but it introduces 2 warnings on PPC64LE specifically:

```
core_dynrec/../../fpu/fpu_instructions.h:145:49: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
```
When writing a fix I noticed a problematic code around this line. I need to do more investigation/testing, as it might (not confirmed yet) indicate a significant bug on big-endian machines (if my suspicions are correct, the bug would result in broken floating-point calculations in some contexts).

To be addressed in a separate PR; it appears on PPC64LE build, but affects PowerPC and s390x only (big-endian architectures); it is NOT introduced by this PR (exists in code for a loooong time already).

```
core_dynrec/risc_ppc64le.h:127:92: warning: overflow in implicit constant conversion [-Woverflow]
```
This warning mysteriously disappears as soon as I move calculation out of macro. Also, I can't reproduce it using clang… I suspect it might be a bug in GCC, as the warning is reported in some contexts but not in others (where I would expect it as well). There's an easy trick to prevent it from showing up, but I want to be 100% sure there's no bug in GCC first.

To be addressed later; affects PPC64LE only (if there's really a problem).

kcgen commented:
PPC optimized build is grinding away; will report back after testing.

kcgen commented:
Passed everything I threw at it on the PPC, Phil's benchmark pack, Doom, Quake, MK3, Stunts, old and new Sierra titles.  Also test CD-DA (piles of math in the codecs; albeit interpreted by the host and not the dynrec engine), and all is working well.

dreamer commented:
@kcgen merging this in then - it'll unblock us to bring back alignment-safe functions and start `cache.h` cleanup :)

dreamer commented:
Merged-in wth f333036d5619086605e5d113e83b33ee06a6c2df

kcgen commented:
> it'll unblock us to bring back alignment-safe functions and start cache.h cleanup :)

Cheers to that @dreamer.  Thanks for tackling this huge merge effort!
